client/ayon_core/pipeline/workfile/workfile_template_builder.py

```diff
@@ -533,29 +533,20 @@ def build_template(
         if create_first_version:
```

Comment:
Like I suggested [here](https://github.com/ynput/ayon-houdini/pull/74#issuecomment-2331975304), the code is more readable to me like this:
```python
def build_template(
    self,
    template_path=None,
    level_limit=None,
    keep_placeholders=None,
    create_first_version=None,
    workfile_creation_enabled=False
):
    # Get default values if not provided
    if template_path is None or keep_placeholders is None or create_first_version is None:
        preset = self.get_template_preset()
        template_path = template_path or preset["path"]
        keep_placeholders = keep_placeholders if keep_placeholders is not None else preset["keep_placeholder"]
        create_first_version = create_first_version if create_first_version is not None else preset["create_first_version"]

    created_version_workfile = self.create_first_workfile_version() if create_first_version else False

    # Handle workfile creation if enabled and a new version was created
    if workfile_creation_enabled and created_version_workfile:
        self.import_template(template_path)
        self.populate_scene_placeholders(level_limit, keep_placeholders)
        self.save_workfile(created_version_workfile)
        return

    # If workfile creation is not enabled or no new version was created,
    # proceed with template import and population
    if not workfile_creation_enabled:
        self.import_template(template_path)
        self.populate_scene_placeholders(level_limit, keep_placeholders)
```

Reply:
can you expand why? is it because otherwise the template import/populate runs when scene is not empty?

Reply:
> Please note that your logic misses `self.host.get_current_workfile() is None`.. from my testing, it's essential.

I don't think that's foolproof currently.

I think some integrations _unfortunately_ may return `None` or even `"untitled.hip"` (houdini?) depending on what the software returns as 'current filename' - even in an empty scene.

**Also** I'd argue that the logic to decide that this runs only on a new EMPTY scene should potentially be outside of this? Like, allow the host integration to define whether it opened something or not and then decide whether to run the template build to begin with?

Reply:
> I think some integrations _unfortunately_ may return `None` or even `"untitled.hip"` (houdini?) depending on what the software returns as 'current filename' - even in an empty scene.

I checked Maya, Nuke and Houdini they all return None. 

> **Also** I'd argue that the logic to decide that this runs only on a new EMPTY scene should potentially be outside of this? Like, allow the host integration to define whether it opened something or not and then decide whether to run the template build to begin with?

tbh, I don't know.

--- 

tbh, figuring out how this feature should work is not easy for me. 
I may/might be adding my personal opinion in this PR but by looking at the available settings and reading through your discussions made me come up with the following user story: 
1. if a workfile has a matched template profile, that profile will be picked.
2. the workfile will be built on any empty scene no matter there are existing files or not. 
3. if there are no existing files and the option `create first version` is enabled, then the work file will be saved.

and, the current state of this PR at this commit https://github.com/ynput/ayon-core/pull/871/commits/e4881e2fec7599863e311fd9124db89bc35fd2b8 exactly does what I've just described. but I still think the code can be furtherly improved.

Reply:
Good catch, 
so having this profile 
![image](https://github.com/user-attachments/assets/aa00e4b2-3142-4ce9-8bab-8b83d0e5acc2)
leads to this error and _no files are created._
```python
Traceback (most recent call last):
  File "C:\PROGRA~1/SIDEEF~1/HOUDIN~1.805/houdini/python3.9libs\hdefereval.py", line 155, in _processDeferred
    result = code(*args, **kwargs)
  File "E:\Ynput\ayon-houdini\client\ayon_houdini\api\lib.py", line 1404, in start_workfile_template_builder
    build_workfile_template(workfile_creation_enabled=True)
  File "E:\Ynput\ayon-houdini\client\ayon_houdini\api\workfile_template_builder.py", line 245, in build_workfile_template
    builder.build_template(*args, **kwargs)
  File "E:\Ynput\ayon-core\client\ayon_core\pipeline\workfile\workfile_template_builder.py", line 517, in build_template
    preset = self.get_template_preset()
  File "E:\Ynput\ayon-houdini\client\ayon_houdini\api\workfile_template_builder.py", line 83, in get_template_preset
    raise TemplateLoadFailed((
ayon_core.pipeline.workfile.workfile_template_builder.TemplateLoadFailed: Template path is not set.
Path need to be set in Houdini\Template Workfile Build Settings\Profiles
```

Reply:
that error makes sense to me if we assume that `Create first version` it's associated to the Workfile template build framework and not outside of it (which is the case given how the settings are shown). For it to not error out it should be part of a generic `Workfile` settings where workfiles (even if empty) are created on launch

Reply:
> I've asked now says "yes, make sense for this to work to save a first version even without a template" so if it's easier to update the code now to allow that.

going quickly through the code, template builder always expect a template to be set.
as there many places that will/may error if template path is empty or None.  
e.g. `open_template`

https://github.com/ynput/ayon-core/blob/167ec7a3ee6bf0b239b9f4115f4997ea251e2fa1/client/ayon_core/pipeline/workfile/workfile_template_builder.py#L590-L594

Reply:
> therefore, I think decoupling building the template from saving the workfile needs another PR..

Perfect.